pkg/api/types.go

```diff
@@ -688,36 +691,66 @@ type ServiceStatus struct{}
 
 // ServiceSpec describes the attributes that a user creates on a service
 type ServiceSpec struct {
-	// Port is the TCP or UDP port that will be made available to each pod for connecting to the pods
-	// proxied by this service.
-	Port int `json:"port"`
+	// Required: The list of ports that are exposed.
+	// FIXME: this would be nicer as a map of name -> struct
+	Ports []ServicePort
 
-	// Optional: Supports "TCP" and "UDP".  Defaults to "TCP".
-	Protocol Protocol `json:"protocol,omitempty"`
+	// Either 'Selector' or 'Endpoints' must be set:
 
-	// This service will route traffic to pods having labels matching this selector. If empty or not present,
-	// the service is assumed to have endpoints set by an external process and Kubernetes will not modify
-	// those endpoints.
+	// This service will route traffic to pods having labels matching this
+	// selector.
 	Selector map[string]string `json:"selector"`
 
+	// Use the specified endpoints.
+	// FIXME: Some weird interaction.  All the fields of Ports[] are basically
+	//   ignored in favor of Endpoints.  Maybe make this just a list of IPs,
+	//   and disallow named DestinationPorts if using this mode?
+	Endpoints []Endpoint
+
 	// PortalIP is usually assigned by the master.  If specified by the user
 	// we will try to respect it or else fail the request.  This field can
 	// not be changed by updates.
 	PortalIP string `json:"portalIP,omitempty"`
 
-	// CreateExternalLoadBalancer indicates whether a load balancer should be created for this service.
+	//FIXME: Should these next 3 be per-port (different external IPs?) or for all ports?
+	// CreateExternalLoadBalancer indicates whether a load balancer should
+	// be created for this service.
 	CreateExternalLoadBalancer bool `json:"createExternalLoadBalancer,omitempty"`
+
 	// PublicIPs are used by external load balancers.
 	PublicIPs []string `json:"publicIPs,omitempty"`
 
-	// ContainerPort is the name of the port on the container to direct traffic to.
-	// Optional, if unspecified use the first port on the container.
-	ContainerPort util.IntOrString `json:"containerPort,omitempty"`
-
-	// Optional: Supports "ClientIP" and "None".  Used to maintain session affinity.
+	// Optional: Supports "ClientIP" and "None".  Used to maintain session
+	// affinity.
 	SessionAffinity AffinityType `json:"sessionAffinity,omitempty"`
 }
 
+type ServicePort struct {
+	// Optional if only one ServicePort is defined on this service,
+	// otherwise required: The name of this port within the service.  This
+	// must be a DNS_LABEL.  All ports within a ServiceSpec must have
+	// unique roles.  This maps to the 'Role' field in Endpoint objects.
+	Role string
+
+	// Optional: The IP protocol for this port.  Supports "TCP" and "UDP".
+	// Defaults to "TCP".
+	Protocol Protocol
+
+	// Required: The port that will be exposed by this service.
+	Port int
+
+	// Optional: How many contiguous ports to expose.  If this is not
```

Comment:
If you have two single ports, you can use two ServicePort definitions.  If
you have a range you can use a range.  Or you can mix and match.  Maybe I
am not understanding the question?

On Wed, Feb 11, 2015 at 10:57 PM, roberthbailey <notifications@github.com>
wrote:

> In pkg/api/types.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4370#discussion_r24561970>
> :
>
> >  	SessionAffinity AffinityType `json:"sessionAffinity,omitempty"`
> >  }
> >
> > +type ServicePort struct {
> > +	// Optional if only one ServicePort is defined on this service,
> > +	// otherwise required: The name of this port within the service.  This
> > +	// must be a DNS_LABEL.  All ports within a ServiceSpec must have
> > +	// unique roles.  This maps to the 'Role' field in Endpoint objects.
> > +	Role string
> > +
> > +	// Optional: The IP protocol for this port.  Supports "TCP" and "UDP".
> > +	// Defaults to "TCP".
> > +	Protocol Protocol
> > +
> > +	// Required: The port that will be exposed by this service.
> > +	Port int
> > +
> > +	// Optional: How many contiguous ports to expose.  If this is not
>
> Are the use cases we've seen for needing to expose multiple ports in a
> service covered by contiguous ports? It seems like this could lead to
> opening a huge block of ports that are unnecessary between the ports that
> users care about.
>
> For instance, the Zookeeper getting started guide recommends using ports
> 2888 & 3888 for the service (
> http://zookeeper.apache.org/doc/r3.1.2/zookeeperStarted.html#sc_RunningReplicatedZooKeeper
> ):
>
> tickTime=2000
> dataDir=/var/zookeeper
> clientPort=2181
> initLimit=5
> syncLimit=2
> server.1=zoo1:2888:3888
> server.2=zoo2:2888:3888
> server.3=zoo3:2888:3888
>
> Which would mean you'd need to open 1000 contiguous ports here to make it
> work with the recommended replicated configuration.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4370/files#r24561970>
> .
>